sdk/couchbase/src/collection_ds.rs

```diff
@@ -0,0 +1,444 @@
+            .await?;
+
+        res.content_as(0)
+    }
+
+    pub async fn insert<V: Serialize>(
+        &self,
+        id: impl Into<String>,
+        value: V,
+    ) -> crate::error::Result<()> {
+        self.collection
+            .mutate_in(
+                &self.id,
+                &[MutateInSpec::upsert(id, value, None)?],
+                MutateInOptions::builder()
+                    .store_semantics(StoreSemantics::Upsert)
+                    .build(),
+            )
+            .await?;
+
+        Ok(())
+    }
+
+    pub async fn remove(&self, id: impl Into<String>) -> crate::error::Result<()> {
+        self.collection
+            .mutate_in(&self.id, &[MutateInSpec::remove(id, None)], None)
+            .await?;
+
+        Ok(())
+    }
+
+    pub async fn contains_key(&self, id: impl Into<String>) -> crate::error::Result<bool> {
+        let res = self
+            .collection
+            .lookup_in(&self.id, &[LookupInSpec::exists(id, None)], None)
+            .await?;
+
+        res.exists(0)
+    }
+
+    pub async fn len(&self) -> crate::error::Result<usize> {
+        let res = self
+            .collection
+            .lookup_in(&self.id, &[LookupInSpec::count("", None)], None)
+            .await?;
+
+        res.content_as(0)
+    }
+
+    pub async fn keys(&self) -> crate::error::Result<Vec<String>> {
+        let res = self.collection.get(&self.id, None).await?;
+
+        let map_contents: HashMap<String, serde_json::Value> = res.content_as()?;
+        Ok(map_contents.keys().cloned().collect())
+    }
+
+    pub async fn values<T: DeserializeOwned>(&self) -> crate::error::Result<Vec<T>> {
+        let res = self.collection.get(&self.id, None).await?;
+
+        let map_contents: HashMap<String, T> = res.content_as()?;
+        Ok(map_contents.into_values().collect())
+    }
+
+    pub async fn clear(&self) -> crate::error::Result<()> {
+        self.collection.remove(&self.id, None).await?;
+        Ok(())
+    }
+}
+
+#[derive(Clone)]
+pub struct CouchbaseSet<'a> {
+    pub collection: &'a Collection,
+    pub id: String,
+    pub options: CouchbaseSetOptions,
+}
+
+impl CouchbaseSet<'_> {
+    pub async fn iter<T: DeserializeOwned>(&self) -> crate::error::Result<impl Iterator<Item = T>> {
+        let res = self.collection.get(&self.id, None).await?;
+        let list_contents: Vec<T> = res.content_as()?;
+
+        Ok(list_contents.into_iter())
+    }
+
+    pub async fn insert<V: Serialize>(&self, value: V) -> crate::error::Result<(bool)> {
+        let res = self
+            .collection
+            .mutate_in(
+                &self.id,
+                &[MutateInSpec::array_add_unique("", value, None)?],
+                MutateInOptions::builder()
+                    .store_semantics(StoreSemantics::Upsert)
+                    .build(),
+            )
+            .await;
+
+        if let Err(e) = res {
+            return if e.msg.contains("subdoc path exists") {
```

Comment:
Do we not have a way to detect these errors without string parsing? If not we should probably add a TODO here to fix this when we add that.

Reply:
Done